samples/nrf9160/modem_shell/src/fota/fota.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr.h>
+#include <power/reboot.h>
+#include <shell/shell.h>
+#include <modem/at_cmd.h>
+#include <net/fota_download.h>
+
+#include "fota.h"
+
+#define MOSH_FOTA_TLS_SECURITY_TAG 4242424
+
+#define AT_CMNG_TYPE_ROOT_CA_CERT 0
+#define AT_CMNG_TYPE_CLIENT_CERT 1
+
+static const char root_ca_cert[] = {
+#include "cert/Baltimore-CyberTrust-Root"
+};
+
+static const char at_cmng_list_template[] = "AT%%CMNG=1,%d,%d";
+static const char at_cmng_write_template[] = "AT%%CMNG=0,%d,%d,\"%s\"";
+
+extern const struct shell *shell_global;
+
+static void reboot_timer_handler(struct k_timer *dummy)
+{
+	sys_reboot(SYS_REBOOT_WARM);
+}
+
+K_TIMER_DEFINE(reboot_timer, reboot_timer_handler, NULL);
+
+static const char *get_error_cause(enum fota_download_error_cause cause)
+{
+	switch (cause) {
+	case FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED:
+		return "download failed";
+	case FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE:
+		return "invalid update";
+	default:
+		return "unknown cause value";
+	}
+}
+
+static void fota_download_callback(const struct fota_download_evt *evt)
+{
+	switch (evt->id) {
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		shell_print(shell_global, "FOTA: Progress %d%%",
+			    evt->progress);
+		break;
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		shell_print(
+			shell_global,
+			"FOTA: Download finished, rebooting in 5 seconds...");
+		k_timer_start(&reboot_timer, K_SECONDS(5), K_NO_WAIT);
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		shell_print(shell_global, "FOTA: Still erasing...");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		shell_print(shell_global, "FOTA: Erasing finished");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		shell_error(shell_global, "FOTA: Error, %s",
+			    get_error_cause(evt->cause));
+		break;
+	default:
+		shell_error(shell_global, "FOTA: Unknown event %d",
+			    evt->id);
+		break;
+	}
+}
+
+static bool fota_ca_cert_exists(void)
+{
+	int err;
+	char buf[128];
+
+	sprintf(buf, at_cmng_list_template, MOSH_FOTA_TLS_SECURITY_TAG,
+		AT_CMNG_TYPE_ROOT_CA_CERT);
+	err = at_cmd_write(buf, buf, sizeof(buf), NULL);
+	if (err) {
+		return false;
+	}
+
+	return strlen(buf) == 0 ? false : true;
+}
+
+static int fota_write_ca_cert(void)
+{
+	int err;
+	char *buf;
+
+	err = 0;
+
+	buf = k_malloc(sizeof(root_ca_cert) + 32);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	sprintf(buf, at_cmng_write_template, MOSH_FOTA_TLS_SECURITY_TAG,
+		AT_CMNG_TYPE_ROOT_CA_CERT, root_ca_cert);
+	err = at_cmd_write(buf, NULL, 0, NULL);
```

Comment:
I changed the implementation to use modem_key_mgmt_write() instead and noticed that there's a bug. It enables AT+CMEE to get more detailed error information before writing the certificate to the modem, but function write_at_cmd_with_cme_enabled() has a bug because of which the actual err from at_cmd_write() gets lost. So now when I call the modem_key_mgmt_write() writing of the certificate fails (I checked from modem log), but the function returns 0. I don't yet know why it even fails, it works with the same data using AT commands.

Reply:
Created bug NCSDK-9783 for this issue.

Reply:
modem_key_mgmt_write() is broken, I don't think it can ever write a certificate successfully to the modem. The constructed AT command is not right, it's missing the terminating double quote. I will update the bug so that both problems get fixed, but I'll keep my AT command based implementation for now because that works. :)

Reply:
Ok, in actually works if the length of the certificate is given with strlen() instead of sizeof(), but only because the scratch buffer is initialized to zero on the first run. This really needs some rework. I updated the bug with the issues I found.

Reply:
That was fast, there's already a PR to fix the problems: https://github.com/nrfconnect/sdk-nrf/pull/4632

I will commit the changes to use modem_key_mgmt_*() instead of AT commands.